File: android/common/compose/src/main/java/com/gchristov/newsfeed/android/common/compose/elements/list/AppGroupedList.kt
package com.gchristov.newsfeed.android.common.compose.elements.list

import android.os.Parcelable
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.*
import androidx.compose.material.Divider
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import com.gchristov.newsfeed.android.common.compose.elements.AppText
import com.gchristov.newsfeed.android.common.compose.theme.Theme
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.RawValue

@Composable
fun AppGroupedList(
    modifier: Modifier = Modifier,
    state: LazyListState = rememberLazyListState(),
    showDividers: Boolean = true,
    content: AppGroupedListScope.() -> Unit
) {
    LazyColumn(
        modifier = modifier.fillMaxSize(),
        state = state,
        contentPadding = PaddingValues(AppGroupedListSpacing)
    ) {
        RealAppGroupedListScope(
            lazyListScope = this,
            showDividers = showDividers
        ).content()
    }
}

@LazyScopeMarker
interface AppGroupedListScope {
    fun group(
        key: Any,
        header: @Composable (() -> String)? = null,
        content: AppGroupScope.() -> Unit
    )

    fun item(
        key: Any,
        content: @Composable () -> Unit
    )
}

@LazyScopeMarker
interface AppGroupScope {
    val groupKey: Any

    fun items(
        count: Int,
        key: (index: Int) -> Any,
        itemContent: @Composable LazyItemScope.(index: Int) -> Unit
    )
}

@Suppress("unused")
inline fun <T> AppGroupScope.items(
    items: List<T>,
    noinline key: (item: T) -> Any,
    crossinline itemContent: @Composable LazyItemScope.(item: T) -> Unit
) = items(items.size, { index: Int -> key(items[index]) }) {
    itemContent(items[it])
}

private class RealAppGroupedListScope(
    private val lazyListScope: LazyListScope,
    private val showDividers: Boolean
) : AppGroupedListScope {
    private var groupCount = 0

    override fun group(
        key: Any,
        header: @Composable (() -> String)?,
        content: AppGroupScope.() -> Unit
    ) {
        val isFirstGroup = groupCount == 0
        groupCount++

        lazyListScope.item(key = key) {
            if (header != null) {
                var topPadding = AppGroupedListSpacing / 2
                if (!isFirstGroup) {
                    topPadding += AppGroupedListSpacing
                }
                AppText(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(
                            top = topPadding,
                            start = AppGroupedListSpacing,
                            bottom = 12.dp,
                            end = AppGroupedListSpacing
                        ),
                    text = header(),
                    style = Theme.typography.subtitle,
                    color = Theme.contentColors.secondary
                )
            } else if (!isFirstGroup) {
                // If there's no header ensure there's still some spacing between the sections
                // except the first one
                Spacer(modifier = Modifier.size(AppGroupedListSpacing))
            }
        }
        RealAppGroupScope(
            groupKey = key,
            lazyListScope = lazyListScope,
            showDividers = showDividers
        ).apply(content)
    }

    override fun item(
        key: Any,
        content: @Composable () -> Unit
    ) {
        lazyListScope.item(key = key) {
            content()
        }
    }
}

private class RealAppGroupScope(
    override val groupKey: Any,
    private val lazyListScope: LazyListScope,
    private val showDividers: Boolean
) : AppGroupScope {
    private var itemCount = 0

    override fun items(
        count: Int,
        key: (index: Int) -> Any,
        itemContent: @Composable LazyItemScope.(index: Int) -> Unit
    ) {
        val firstItemIndex = itemCount
        itemCount += count

        lazyListScope.items(
            count = count,
            key = { index -> AppCompositeKey(groupKey, key(index)) }
        ) { index ->
            this@RealAppGroupScope.Item(itemIndex = firstItemIndex + index) { itemContent(index) }
        }
    }

    @Composable
    private fun Item(
        itemIndex: Int,
        content: @Composable () -> Unit
    ) {
        val isAtLastRow = itemIndex == itemCount - 1
        val shape = when {
            itemCount == 1 -> Theme.shapes.groupSingle
            itemIndex == 0 -> Theme.shapes.groupStart
            isAtLastRow -> Theme.shapes.groupEnd
            else -> Theme.shapes.groupMiddle
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(shape)
                .background(Theme.backgrounds.surface)
        ) {
            content()

            if (showDividers && !isAtLastRow) {
                Divider(color = Theme.backgrounds.primary)
            }
        }
    }
}

@Parcelize
private data class AppCompositeKey(
    val groupKey: @RawValue Any,
    val itemKey: @RawValue Any
) : Parcelable

private val AppGroupedListSpacing = 16.dp

File: ios/common-swiftui/common-swiftui/elements/list/AppGroupedList.swift
import SwiftUI

public struct AppGroupedList<Content>: View where Content: View {
    private let showDividers: Bool
    private let content: (AppGroupedListScope) -> Content
    
    public init(
        showDividers: Bool = true,
        @ViewBuilder content: @escaping (AppGroupedListScope) -> Content
    ) {
        self.showDividers = showDividers
        self.content = content
        // Remove list dividers from UIKit views
        UITableView.appearance().separatorColor = .clear
    }
    
    public var body: some View {
        List {
            content(AppGroupedListScope(showDividers: showDividers))
        }
        .modifier(AppListStyle())
    }
}

public class AppGroupedListScope {
    private let showDividers: Bool
    private var groupCount = 0
    
    init(showDividers: Bool) {
        self.showDividers = showDividers
    }
    
    public func group<Content: View>(
        header: String? = nil,
        @ViewBuilder content: (AppGroupScope) -> Content
    ) -> some View {
        let isFirstGroup = groupCount == 0
        let horizontalPadding = AppGroupedListSpacing * 2
        var topPadding = AppGroupedListSpacing / 2
        if (!isFirstGroup) {
            topPadding += AppGroupedListSpacing
        }
        groupCount += 1
        
        return Group {
            AppGroupHeader(text: header).padding(EdgeInsets(top: topPadding, leading: horizontalPadding, bottom: 12, trailing: horizontalPadding))
            content(AppGroupScope(showDividers: showDividers))
        }
    }
    
    public func item<Content: View>(@ViewBuilder content: () -> Content) -> some View {
        content()
    }
}

public class AppGroupScope {
    private let showDividers: Bool
    private var itemCount = 0
    
    init(showDividers: Bool) {
        self.showDividers = showDividers
    }
    
    public func items<Data: AnyObject, Content: View>(
        items: [Data],
        @ViewBuilder itemContent: @escaping (Int) -> Content
    ) -> some View {
        let firstItemIndex = itemCount
        itemCount += items.count
        
        return ForEach(0 ..< items.count, id: \.self) { index in
            self.item(
                itemIndex: firstItemIndex + index,
                content: itemContent
            )
        }
    }
    
    private func item<Content: View>(
        itemIndex: Int,
        @ViewBuilder content: @escaping (Int) -> Content
    ) -> some View {
        return AppGroupItem(
            showDivider: showDividers,
            itemIndex: itemIndex,
            itemCount: itemCount,
            content: content
        )
    }
}

private struct AppGroupHeader: View {
    @EnvironmentObject var theme: Theme
    private let text: String?
    
    init(text: String?) {
        self.text = text
    }
    
    public var body: some View {
        ZStack {
            if let text = text {
                AppText(
                    text: text.uppercased(),
                    color: theme.contentColors.secondary,
                    font: theme.typography.subtitle
                )
            }
        }
        .modifier(AppListItemStyle())
    }
}

private struct AppGroupItem<Content>: View where Content: View {
    @EnvironmentObject var theme: Theme
    private let showDivider: Bool
    private let itemIndex: Int
    private let itemCount: Int
    private let isAtLastItem: Bool
    private let content: (Int) -> Content
    
    init(
        showDivider: Bool,
        itemIndex: Int,
        itemCount: Int,
        @ViewBuilder content: @escaping (Int) -> Content
    ) {
        self.showDivider = showDivider
        self.itemIndex = itemIndex
        self.itemCount = itemCount
        self.content = content
        self.isAtLastItem = itemIndex == itemCount - 1
    }
    
    private func shapeForPosition(
        itemIndex: Int,
        itemCount: Int,
        isAtLastItem: Bool
    ) -> some Shape {
        if itemCount == 1 {
            return theme.shapes.groupSingle
        } else if itemIndex == 0 {
            return theme.shapes.groupStart
        } else if isAtLastItem {
            return theme.shapes.groupEnd
        }
        return theme.shapes.groupMiddle
    }
    
    public var body: some View {
        VStack(spacing: 0) {
            content(itemIndex)
            if (showDivider && !isAtLastItem) {
                Rectangle()
                    .fill(theme.backgrounds.primary)
                    .frame(height: 1)
            }
        }
            .background(theme.backgrounds.surface)
            .clipShape(shapeForPosition(
                itemIndex: itemIndex,
                itemCount: itemCount,
                isAtLastItem: isAtLastItem
            ))
            .padding(.horizontal, AppGroupedListSpacing)
            .modifier(AppListItemStyle())
    }
}

private let AppGroupedListSpacing = 16.0
